compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2239,7 +2239,9 @@ impl<'a> Parser<'a> {
 
     /// Parses the condition of a `if` or `while` expression.
     fn parse_cond_expr(&mut self) -> PResult<'a, P<Expr>> {
+        self.is_in_a_valid_let_ctx = true;
         let cond = self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None)?;
+        self.is_in_a_valid_let_ctx = false;
```

Comment:
Rename `is_in_a_valid_let_ctx` to `let_expr_allowed`. Then can you pull this into a function, like:

``` rust
fn with_let_expr_allowed(&mut self, allowed: bool, f: impl FnOnce(&mut Self) -> T) -> T {
    let last_let_expr_allowed = std::mem::replace(&mut self.let_expr_allowed, allowed);
    f(self);
    self.let_expr_allowed = last_let_expr_allowed;
}
```

Then do this:
```suggestion
        let cond = self.with_let_expr_allowed(allowed, |self_|  self_.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None))?;
```